### PR TITLE
Simplify `telegram::Client` API

### PR DESCRIPTION
**Pull request by hcpl**
`futures`, `tokio` and `hyper` usages are implementation details.

### CODE REVIEW

**Comment by mehcode** [comment]
@hcpl Don't have time to do a full review but I don't agree with several changes. I'll review it in a few hours.

File: examples/create_auth_key/src/main.rs [hunk 1]
```diff
@@ -35,16 +33,14 @@ fn main() {
     // [DEBUG] Step
     println!(" - Send {}\n", "http://149.154.167.50:443/api");
 
-    let mut core = Core::new().unwrap();
-    let client = telegram::Client::new(&core.handle());
-    let promise = client.request(req).map(|data| {
+    let mut client = telegram::Client::new()?;
```
Comment (mehcode) [inline examples/create_auth_key/src/main.rs]: From what I understand about `tokio`, the event loops is the `Core`, and I/O operations must take a `Handle` or `Remote` to schedule their work.

I poked around and the 2 1/2 tokio web libraries/frameworks so far don't seem to understand how an event loop interoperates. They keep it internal and don't provide a way to schedule work. There should be a single event loop per thread. An event loop is a way of "sharing" CPU time while waiting for I/O to finish. If you create N event loops you lose all of the benefits.

If I was working with several libraries, I'd accept to be in charge of creating a `Core` and passing handles to it where needed and then at the bottom of `main`, calling `core.run( .. )`

See [postgres](https://github.com/sfackler/rust-postgres/blob/d0db938f825ecea5ba2bb26d08df4db9a5589562/tokio-postgres/src/lib.rs#L26), [rpc](https://github.com/iorust/tokio-rpc/blob/master/examples/protobuf_rpc.rs#L67), or [kinetic](https://github.com/peterdelevoryas/tokio-kinetic/blob/master/examples/client.rs#L46) for some examples of async clients.

File: telegram/src/client.rs [hunk 2]
```diff
@@ -1,40 +1,51 @@
...
-    pub fn request<T: Serialize>(
-        &self,
-        req: Request<T>,
-    ) -> Box<Future<Item = Vec<u8>, Error = error::Error>> {
-        let http_request = match req.to_http_request() {
-            Ok(req) => req,
-            Err(error) => return Box::new(future::err(error)),
-        };
+    /// Send a constructed request using this Client.
+    pub fn send<F, T, U>(&mut self, req: Request<T>, on_receive_handler: F) -> error::Result<U>
```
Comment (mehcode) [inline telegram/src/client.rs]: More on interoperability. Everything in async Rust consumes Futures and returns Futures. From what I understand the result of [this RFC] will mean `Future` will become a trait in `std`.

File: examples/create_auth_key/src/main.rs (hunk 1 above)
Comment (hcpl) [inline examples/create_auth_key/src/main.rs]: I agree that performance is a great deal in a language with zero-cost features like Rust. But with the current state of Rust's futures I guess there are users who don't want to trade code conciseness for runtime efficiency. I don't know if they are right or wrong about "readability matters" in a performance-oriented language, I just think there is a demand for simple APIs based on the sole existence of `reqwest`.

On the other hand, I don't have enough experience with async-oriented API design anyway (not counting non-`futures` Rust, I've only used pre-3.5 Python, QML flavor of JS and JS in context of simple DOM-modifying userscripts - and those are languages with extensive support for async for a while now). So I'd be glad to learn more about async in general!

File: telegram/src/client.rs (hunk 2 above)
Comment (hcpl) [inline telegram/src/client.rs]: I guess the RFC you wanted to refer is [this one](https://github.com/rust-lang/rfcs/issues/1081).

**Pull request closed without merging**